console-subscriber/src/aggregator.rs

```diff
@@ -615,9 +906,287 @@ impl Task {
+                } as i32;
+                PbAttrUpdate {
+                    name: self.name.clone(),
+                    update: Some(PbUpdate::Numeric(PbNumeric { val, op, unit })),
+                }
+            }
+        }
+    }
+}
+
+impl ToProto for AttrValue {
+    type Result = proto::resources::stats::AttrValue;
+
+    fn to_proto(&self) -> Self::Result {
+        use proto::resources::stats::attr_value::Numeric;
+        use proto::resources::stats::attr_value::Val;
+        use proto::resources::stats::AttrValue as PbAttrValue;
+
+        match self {
+            AttrValue::Text(t) => PbAttrValue {
+                val: Some(Val::Text(t.clone())),
+            },
+            AttrValue::Numeric { val, unit } => PbAttrValue {
+                val: Some(Val::Numeric(Numeric {
+                    val: *val,
+                    unit: unit.clone(),
+                })),
+            },
+        }
+    }
+}
+
+impl AttrValue {
+    fn update(&mut self, update: &AttributeUpdateValue) {
+        match (self, update) {
+            (AttrValue::Text(t), AttributeUpdateValue::Text(upd)) => {
+                *t = upd.clone();
+            }
+            (
+                AttrValue::Numeric { ref mut val, .. },
+                AttributeUpdateValue::Numeric { val: upd, op, .. },
+            ) => match op {
+                AttributeUpdateOp::Add => {
+                    let new_val = *val + upd;
+                    *val = new_val;
+                }
+                AttributeUpdateOp::Ovr => {
+                    *val = *upd;
+                }
+                AttributeUpdateOp::Sub => {
+                    let new_val = *val - upd;
+                    *val = new_val;
+                }
+            },
+            _ => panic!("trying to update attributes of different type"),
+        }
+    }
+}
+
+impl From<AttributeUpdateValue> for AttrValue {
+    fn from(upd: AttributeUpdateValue) -> Self {
+        match upd {
+            AttributeUpdateValue::Text(t) => AttrValue::Text(t),
+            AttributeUpdateValue::Numeric { val, unit, .. } => AttrValue::Numeric { val, unit },
+        }
+    }
+}
+
 fn serialize_histogram(histogram: &Histogram<u64>) -> Result<Vec<u8>, V2SerializeError> {
     let mut serializer = V2Serializer::new();
     let mut buf = Vec::new();
     serializer.serialize(histogram, &mut buf)?;
     Ok(buf)
 }
+
+fn total_time(created_at: Option<&SystemTime>, closed_at: Option<&SystemTime>) -> Option<Duration> {
+    closed_at.and_then(|end| created_at.and_then(|start| end.duration_since(*start).ok()))
+}
+
+/// Drops all tasks, resources and ops that are not alive anymore
+fn drop_closed<T, R: Closable>(
+    now: SystemTime,
+    entities: &mut IdData<T>,
+    stats: &mut IdData<R>,
+    retention: Duration,
+    has_watchers: bool,
+) {
+    // drop stats for closed tasks if they have been updated
+    tracing::trace!(?retention, has_watchers, "dropping closed entities...");
```

Comment:
nit/tioli: maybe add something like 
```suggestion
    tracing::trace!(?retention, has_watchers, "dropping closed {}...". std::any::type_name::<T>());
```
since we'll now log this message three times in a row and it isn't clear whether we're dropping tasks, resources, or async ops...